rust/src/origin.rs

```diff
@@ -6,56 +6,236 @@
+
+    let mut cfg: crate::treefile::TreeComposeConfig = Default::default();
+    let refspec_str = if let Some(r) = keyfile_get_optional_string(&kf, ORIGIN, "refspec")? {
+        Some(r)
+    } else {
+        keyfile_get_optional_string(&kf, ORIGIN, "baserefspec")?
+    };
+    cfg.derive.from = refspec_str;
+    cfg.packages = parse_stringlist(&kf, PACKAGES, "requested")?;
+    cfg.derive.packages_local = parse_localpkglist(&kf, PACKAGES, "requested-local")?;
+    cfg.derive.override_remove = parse_stringlist(&kf, OVERRIDES, "remove")?;
+    cfg.derive.override_replace_local = parse_localpkglist(&kf, OVERRIDES, "replace-local")?;
+
+    let regenerate_initramfs = kf
+        .get_boolean(RPMOSTREE, "regenerate-initramfs")
+        .unwrap_or_default();
+    let initramfs_etc = parse_stringlist(kf, RPMOSTREE, "initramfs-etc")?;
+    let initramfs_args = parse_stringlist(kf, RPMOSTREE, "initramfs-args")?;
+    if regenerate_initramfs || initramfs_etc.is_some() || initramfs_args.is_some() {
+        let initramfs = crate::treefile::DeriveInitramfs {
+            regenerate: regenerate_initramfs,
+            etc: initramfs_etc,
+            args: initramfs_args,
+        };
+        cfg.derive.initramfs = Some(initramfs);
+    }
+
+    if let Some(url) = keyfile_get_optional_string(&kf, ORIGIN, "custom-url")? {
+        let description = keyfile_get_optional_string(&kf, ORIGIN, "custom-description")?;
+        cfg.derive.custom = Some(crate::treefile::DeriveCustom { url, description })
+    }
 
-#[derive(Copy, Clone, Debug, PartialEq, Eq)]
-pub(crate) enum RefspecType {
-    Checksum,
-    Ostree,
+    if map_keyfile_optional(kf.get_boolean(RPMOSTREE, "ex-cliwrap"))?.unwrap_or_default() {
+        cfg.cliwrap = Some(true)
+    }
+
+    cfg.derive.override_commit = keyfile_get_optional_string(kf, ORIGIN, "override-commit")?;
+
+    Ok(Box::new(Treefile::new_from_config(cfg, None)?))
 }
 
-#[derive(Clone, Debug, PartialEq, Eq)]
-pub(crate) struct Refspec {
-    kind: RefspecType,
-    value: String,
+/// Convert an origin keyfile to a treefile config.
+///
+/// For historical reasons, rpm-ostree has two file formats to represent
+/// state.  This bridges parts of an origin file to a treefile that
+/// is understood by the core.
+pub(crate) fn origin_to_treefile(
+    mut kf: Pin<&mut crate::ffi::GKeyFile>,
+) -> CxxResult<Box<Treefile>> {
+    let kf = kf.gobj_wrap();
+    Ok(origin_to_treefile_inner(&kf)?)
 }
 
-struct Cache {
-    refspec: Refspec,
-    override_commit: Option<String>,
-    #[allow(dead_code)]
-    unconfigured_state: Option<String>,
+/// Set a keyfile value to a string list.
+fn kf_set_string_list<'a>(
+    kf: &glib::KeyFile,
+    group: impl AsRef<str>,
+    k: impl AsRef<str>,
+    vals: impl IntoIterator<Item = &'a str>,
+) {
+    let mut v = String::new();
+    for elt in vals {
+        v.push_str(elt);
+        v.push(';');
+    }
+    kf.set_value(group.as_ref(), k.as_ref(), v.as_str())
```

Comment:
Minor/optional: can't we use `set_string_list` here to avoid building a `String`?

Reply:
It's not bound right now, I didn't dig into why.